File: dataset/CoT - Copy/OPENAI_gpt-5-2025-08-07/Task61.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

static int is_name_char(char c) {
    return isalnum((unsigned char)c) || c == '_' || c == '-' || c == ':' || c == '.';
}

static size_t skip_spaces(const char* s, size_t i) {
    while (s[i] == ' ' || s[i] == '\t' || s[i] == '\n' || s[i] == '\r') i++;
    return i;
}

static int starts_with(const char* s, size_t i, const char* p) {
    size_t lp = strlen(p);
    size_t ls = strlen(s);
    if (i + lp > ls) return 0;
    return strncmp(s + i, p, lp) == 0;
}

static long index_of_seq2(const char* s, size_t from, char x, char y) {
    size_t n = strlen(s);
    for (size_t i = from; i + 1 < n; ++i) {
        if (s[i] == x && s[i + 1] == y) return (long)(i + 1);
    }
    return -1;
}

static long index_of_seq3(const char* s, size_t from, char x, char y, char z) {
    size_t n = strlen(s);
    for (size_t i = from; i + 2 < n; ++i) {
        if (s[i] == x && s[i + 1] == y && s[i + 2] == z) return (long)(i + 2);
    }
    return -1;
}

static long skip_declaration(const char* s, size_t from) {
    size_t n = strlen(s);
    int depth = 0;
    for (size_t i = from; i < n; ++i) {
        char c = s[i];
        if (c == '[') depth++;
        else if (c == ']') { if (depth > 0) depth--; }
        else if (c == '>' && depth == 0) return (long)i;
    }
    return -1;
}

static long skip_to_tag_end(const char* s, size_t from) {
    size_t n = strlen(s);
    char quote = 0;
    for (size_t i = from; i < n; ++i) {
        char c = s[i];
        if (quote) {
            if (c == quote) quote = 0;
        } else {
            if (c == '\'' || c == '"') quote = c;
            else if (c == '>') return (long)i;
        }
    }
    return -1;
}

char* getRootElementName(const char* xml) {
    if (xml == NULL) {
        char* empty = (char*)calloc(1, 1);
        return empty;
    }
    const char* a = xml;
    size_t n = strlen(a);
    size_t i = 0;

    char* root = NULL;

    while (i < n) {
        const char* ltptr = strchr(a + i, '<');
        if (!ltptr) break;
        i = (size_t)(ltptr - a);
        if (i + 1 >= n) break;

        if (starts_with(a, i, "<?")) {
            long j = index_of_seq2(a, i + 2, '?', '>');
            if (j < 0) break;
            i = (size_t)j + 1;
            continue;
        }

        if (starts_with(a, i, "<!--")) {
            long j = index_of_seq3(a, i + 4, '-', '-', '>');
            if (j < 0) break;
            i = (size_t)j + 1;
            continue;
        }

        if (starts_with(a, i, "<![CDATA[")) {
            long j = index_of_seq3(a, i + 9, ']', ']', '>');
            if (j < 0) break;
            i = (size_t)j + 1;
            continue;
        }

        if (starts_with(a, i, "<!")) {
            long j = skip_declaration(a, i + 2);
            if (j < 0) break;
            i = (size_t)j + 1;
            continue;
        }

        if (a[i + 1] == '/') {
            size_t j = i + 2;
            j = skip_spaces(a, j);
            while (j < n && is_name_char(a[j])) j++;
            long end = skip_to_tag_end(a, j);
            if (end < 0) break;
            i = (size_t)end + 1;
            continue;
        }

        size_t j = i + 1;
        j = skip_spaces(a, j);
        size_t start = j;
        while (j < n && is_name_char(a[j])) j++;
        if (j == start) {
            i++;
            continue;
        }
        size_t len = j - start;
        root = (char*)malloc(len + 1);
        if (!root) {
            // allocation failed
            char* empty = (char*)calloc(1, 1);
            return empty;
        }
        memcpy(root, a + start, len);
        root[len] = '\0';
        return root; // first opening tag is the root
    }

    if (!root) {
        char* empty = (char*)calloc(1, 1);
        return empty;
    }
    return root;
}

int main(void) {
    const char* tests[5] = {
        "<root><child/></root>",
        "<?xml version='1.0'?><!-- comment --><a id='1'><b></b><c/></a>",
        "  \n <root>text<inner>more</inner></root> ",
        "<x><y></x>",
        "<!DOCTYPE note SYSTEM 'Note.dtd'><note><![CDATA[<notatag>]]><to>T</to></note>"
    };
    for (int t = 0; t < 5; ++t) {
        char* r = getRootElementName(tests[t]);
        if (r) {
            printf("%s\n", r);
            free(r);
        } else {
            printf("\n");
        }
    }
    return 0;
}

File: dataset/CoT/OPENAI_gpt-5-2025-08-07/Task61.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

public class Task61 {

    // Node structure to represent XML tree (simplified)
    static class Node {
        String name;
        List<Node> children = new ArrayList<>();
        Node(String name) { this.name = name; }
    }

    // Public API: returns the root element name of the XML document
    public static String getRootElementName(String xml) {
        if (xml == null) return "";
        ParseResult result = parseXML(xml);
        return result.rootName == null ? "" : result.rootName;
    }

    // Internal structure for parse results
    private static class ParseResult {
        String rootName;
        Node rootNode;
    }

    // Parse XML to build a basic structure and identify the root
    private static ParseResult parseXML(String xml) {
        ParseResult res = new ParseResult();
        char[] a = xml.toCharArray();
        int n = a.length;
        int i = 0;

        Deque<Node> stack = new ArrayDeque<>();

        while (i < n) {
            int lt = indexOf(a, '<', i);
            if (lt < 0) break;
            i = lt;

            if (i + 1 >= n) break;

            // Processing instruction
            if (startsWith(a, i, "<?")) {
                int j = indexOfSeq(a, i + 2, '?', '>');
                if (j < 0) break;
                i = j + 1;
                continue;
            }

            // Comment
            if (startsWith(a, i, "<!--")) {
                int j = indexOfSeq(a, i + 4, '-', '-', '>');
                if (j < 0) break;
                i = j + 1;
                continue;
            }

            // CDATA
            if (startsWith(a, i, "<![CDATA[")) {
                int j = indexOfSeq(a, i + 9, ']', ']', '>');
                if (j < 0) break;
                i = j + 1;
                continue;
            }

            // DOCTYPE or other declaration
            if (startsWith(a, i, "<!")) {
                int j = skipDeclaration(a, i + 2);
                if (j < 0) break;
                i = j + 1;
                continue;
            }

            // Closing tag
            if (a[i + 1] == '/') {
                int j = i + 2;
                j = skipSpaces(a, j);
                int start = j;
                while (j < n && isNameChar(a[j])) j++;
                // name = new String(a, start, j - start); // can be used for validation
                j = skipToTagEnd(a, j);
                if (j < 0) break;
                if (!stack.isEmpty()) stack.pop();
                i = j + 1;
                continue;
            }

            // Opening tag
            int j = i + 1;
            j = skipSpaces(a, j);
            int start = j;
            while (j < n && isNameChar(a[j])) j++;
            if (j == start) { // invalid tag name
                i++;
                continue;
            }
            String name = new String(a, start, j - start);
            Node node = new Node(name);
            if (stack.isEmpty()) {
                res.rootName = name;
                res.rootNode = node;
            } else {
                stack.peek().children.add(node);
            }

            // Scan attributes and detect self-close
            boolean selfClose = false;
            int k = j;
            char quote = 0;
            while (k < n) {
                char c = a[k];
                if (quote != 0) {
                    if (c == quote) quote = 0;
                    k++;
                    continue;
                } else {
                    if (c == '\'' || c == '"') {
                        quote = c;
                        k++;
                        continue;
                    }
                    if (c == '/') {
                        if (k + 1 < n && a[k + 1] == '>') {
                            selfClose = true;
                            k += 2;
                            break;
                        }
                    }
                    if (c == '>') {
                        k++;
                        break;
                    }
                    k++;
                }
            }
            if (!selfClose) stack.push(node);
            i = k;
        }

        return res;
    }

    private static boolean startsWith(char[] a, int i, String s) {
        if (i + s.length() > a.length) return false;
        for (int k = 0; k < s.length(); k++) if (a[i + k] != s.charAt(k)) return false;
        return true;
    }

    private static int indexOf(char[] a, char ch, int from) {
        for (int i = Math.max(0, from); i < a.length; i++) if (a[i] == ch) return i;
        return -1;
        return -1;
    }

    // Find sequence x,y,z in order (consecutive)
    private static int indexOfSeq(char[] a, int from, char x, char y, char z) {
        for (int i = Math.max(0, from); i + 2 < a.length; i++) {
            if (a[i] == x && a[i + 1] == y && a[i + 2] == z) return i + 2;
        }
        return -1;
    }

    // For "?>" pair
    private static int indexOfSeq(char[] a, int from, char x, char y) {
        for (int i = Math.max(0, from); i + 1 < a.length; i++) {
            if (a[i] == x && a[i + 1] == y) return i + 1;
        }
        return -1;
    }

    private static int skipSpaces(char[] a, int i) {
        while (i < a.length) {
            char c = a[i];
            if (c == ' ' || c == '\t' || c == '\n' || c == '\r') i++;
            else break;
        }
        return i;
    }

    private static boolean isNameChar(char c) {
        return Character.isLetterOrDigit(c) || c == '_' || c == '-' || c == ':' || c == '.';
    }

    // Skip until matching '>' for declaration (handles internal subset brackets [] depth)
    private static int skipDeclaration(char[] a, int i) {
        int n = a.length;
        int depth = 0;
        for (int k = i; k < n; k++) {
            char c = a[k];
            if (c == '[') depth++;
            else if (c == ']') { if (depth > 0) depth--; }
            else if (c == '>' && depth == 0) return k;
        }
        return -1;
    }

    // Skip to the end of a tag '>' (attributes with quotes handled)
    private static int skipToTagEnd(char[] a, int i) {
        int n = a.length;
        char quote = 0;
        for (int k = i; k < n; k++) {
            char c = a[k];
            if (quote != 0) {
                if (c == quote) quote = 0;
                continue;
            } else {
                if (c == '\'' || c == '"') { quote = c; continue; }
                if (c == '>') return k;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        String[] tests = new String[] {
            "<root><child/></root>",
            "<?xml version='1.0'?><!-- comment --><a id='1'><b></b><c/></a>",
            "  \n <root>text<inner>more</inner></root> ",
            "<x><y></x>",
            "<!DOCTYPE note SYSTEM 'Note.dtd'><note><![CDATA[<notatag>]]><to>T</to></note>"
        };
        for (String t : tests) {
            System.out.println(getRootElementName(t));
        }
    }
}